File: Python/TopoSort_using_dfs.py
from collections import defaultdict

class Graph:
    def __init__(self, vertices):
        self.graph = defaultdict(list)  # Dictionary containing adjacency list
        self.V = vertices  # Number of vertices

    # Function to add an edge to the graph
    def add_edge(self, u, v):
        self.graph[u].append(v)

    # Recursive helper function for DFS and topological sort
    def topological_sort_util(self, v, visited, stack):
        # Mark the current node as visited
        visited[v] = True

        # Recur for all the vertices adjacent to this vertex
        for i in self.graph[v]:
            if not visited[i]:
                self.topological_sort_util(i, visited, stack)

        # Push the current vertex to the stack which stores the result
        stack.append(v)

    # Function to perform topological sort
    def topological_sort(self):
        # Mark all the vertices as not visited
        visited = [False] * self.V
        stack = []

        # Call the recursive helper function to store topological sort
        # starting from all vertices one by one
        for i in range(self.V):
            if not visited[i]:
                self.topological_sort_util(i, visited, stack)

        # Print contents of the stack in reverse order (since it's a stack)
        print(stack[::-1])  # Topological order

# Example usage:
g = Graph(6)
g.add_edge(5, 2)
g.add_edge(5, 0)
g.add_edge(4, 0)
g.add_edge(4, 1)
g.add_edge(2, 3)
g.add_edge(3, 1)

print("Topological Sort of the graph:")
g.topological_sort()

File: Java/TopoSort using dfs.java
import java.util.*;

// Class representing a graph
class Graph {
    private int vertices; // Number of vertices in the graph
    private LinkedList<Integer>[] adjacencyList; // Adjacency list to represent the graph

    // Constructor
    public Graph(int vertices) {
        this.vertices = vertices;
        adjacencyList = new LinkedList[vertices];
        for (int i = 0; i < vertices; i++) {
            adjacencyList[i] = new LinkedList<>();
        }
    }

    // Add an edge to the graph
    public void addEdge(int from, int to) {
        adjacencyList[from].add(to);
    }

    // DFS-based function for Topological Sorting
    private void topologicalSortUtil(int currentVertex, boolean[] visited, Stack<Integer> stack) {
        // Mark the current node as visited
        visited[currentVertex] = true;

        // Recur for all vertices adjacent to this vertex
        for (Integer neighbor : adjacencyList[currentVertex]) {
            if (!visited[neighbor]) {
                topologicalSortUtil(neighbor, visited, stack);
            }
        }

        // Push the current vertex to the stack which stores the result
        stack.push(currentVertex);
    }

    // The function to perform Topological Sort using DFS
    public void topologicalSort() {
        Stack<Integer> stack = new Stack<>(); // Stack to store the topological order
        boolean[] visited = new boolean[vertices]; // Mark all the vertices as not visited

        // Call the recursive helper function for all vertices
        for (int i = 0; i < vertices; i++) {
            if (!visited[i]) {
                topologicalSortUtil(i, visited, stack);
            }
        }

        // Print contents of the stack, which contains the topologically sorted order
        System.out.println("Topological Sort:");
        while (!stack.isEmpty()) {
            System.out.print(stack.pop() + " ");
        }
    }
}

// Driver class to test the code
public class TopoSortUsingDFS {
    public static void main(String[] args) {
        // Create a graph with 6 vertices
        Graph graph = new Graph(6);

        // Add edges to the graph
        graph.addEdge(5, 2);
        graph.addEdge(5, 0);
        graph.addEdge(4, 0);
        graph.addEdge(4, 1);
        graph.addEdge(2, 3);
        graph.addEdge(3, 1);

        // Perform Topological Sort
        graph.topologicalSort();
    }
}
